staging/src/k8s.io/apiserver/pkg/endpoints/handlers/negotiation/negotiate.go

```diff
@@ -280,12 +315,22 @@ func NegotiateMediaTypeOptions(header string, accepted []AcceptedMediaType, endp
 			case clause.Type == accepts.Type && clause.SubType == accepts.SubType,
 				clause.Type == accepts.Type && clause.SubType == "*",
 				clause.Type == "*" && clause.SubType == "*":
-				// TODO: should we prefer the first type with no unrecognized options?  Do we need to ignore unrecognized
-				// parameters.
-				return acceptMediaTypeOptions(clause.Params, accepts, endpoint)
+				candidates = append(candidates, candidateMediaType{accepted: accepts, clauses: clause})
 			}
 		}
 	}
+
+	// run until elements are removed
+	for len(candidates) > 0 {
+		index := mostSpecificMediaType(candidates)
+		if retVal, ret := acceptMediaTypeOptions(candidates[index].clauses.Params, candidates[index].accepted, endpoint); ret {
+			return retVal, true
+		}
+		// swap index with the last item, then truncate the array by one.
+		candidates[index], candidates[len(candidates)-1] = candidates[len(candidates)-1], candidates[index]
```

Comment:
oops, "swap" was incorrect.  You can take the last array item and assign it to the current location.  Then truncate.  

If we don't accept `index`, then we will never accept index and so it doesn't need to be preserved.